README.md

```diff
@@ -143,13 +143,23 @@ password in postgres to `postgres`.
 
 
 ### Run tests
-    (conbench) $ cd ~/workspace/conbench/
-    (conbench) $ pytest -vv conbench/tests/
 
+The canonical test invocation method for now is via `docker-compose` (manages test dependencies automatically, and is also used in CI):
 
-### Run tests with docker
-    (conbench) $ cd ~/workspace/conbench/
-    (conbench) $ docker-compose run app pytest -vv conbench/tests/
+```
+$ docker-compose down && docker-compose build app && \
+    docker-compose run app \
+    pytest -vv conbench/tests
```

Comment:
Do we need both build + run here? I thought run without build would do the building if it needed to?

Reply:
And, is it possible (and advisable) that we mount the local directory so that for tests like the docs tests that will automatically generate the new version (with a warning), will do that + it is persisted in the filesystem?

Reply:
Great question. Yes, the additional `build` is critical to pick up local changes to files.

`run` does not rebuild by itself. I ran into that and found that this is actually expected behavior, documented by docker-compose.

Related: https://github.com/conbench/conbench/issues/437

Reply:
> that we mount the local directory so that for tests like the docs tests that will automatically generate the new version

Mounting a checkout into a test container has typically been the most efficient way I have seen in the past to make a containerized test runner reflect the local checkout on the host (typically faster than a Docker image ADD or COPY).

However, test runner output files then often end up written as `root` onto the host -- that really can be an annoyance. Addressing that can then again increase complexity in an unpleasant way ("do not run processes in container as root", relates to https://docs.docker.com/engine/security/userns-remap/).

Sounds like we need some docs-related output emitted by the test runner every now and then? I'll have to look at the specifics.

Reply:
Ooph, yeah I forgot about the root user thing. In that case, I say let's ship this as is, and we can follow up to see if the docs-related changes can be separated out, or excepted for this case, etc.

Another thought might be hooking up a comment github action that adds the updated docs to the branch when run. That way we can see the failure in CI, then say "@githubbot, please fix up the docs, I did mean to change those" and it would commit the updated expectations? Regardless, IMO that's out of scope for this PR

Reply:
> Another thought might be hooking up a comment github action that adds the updated docs to the branch when run. That way we can see the failure in CI, then say "@githubbot, please fix up the docs, I did mean to change those" and it would commit the updated expectations? Regardless, IMO that's out of scope for this PR

I will have to see what the technical goals are in detail, and then it's easier for me to think about all that.

Maybe this automatic-docs-generation step that we talk about here can be split out of the test suite and run in its own CI step. But that's just me thinking out loud w/o fully understanding what we even talk about :).

In any case, the "current recommended approach to run tests" can of course change again very soon if we feel like  there's a good reason! Personally, I am not the biggest fan of deeply relying on docker-compose.